pkg/rpcmiddleware/validator/validator.go

```diff
@@ -21,11 +21,61 @@ package validator
 import (
 	"context"
 
+	"github.com/gogo/protobuf/types"
 	"go.thethings.network/lorawan-stack/pkg/errors"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 )
 
+var allowedFieldMaskPaths = make(map[string][]string)
+
+// RegisterAllowedFieldMaskPaths registers the allowed field mask paths for an
+// RPC. Note that all allowed paths and sub-paths must be registered.
+func RegisterAllowedFieldMaskPaths(rpcFullMethod string, allowedPaths ...string) {
+	allowedFieldMaskPaths[rpcFullMethod] = append(allowedFieldMaskPaths[rpcFullMethod], allowedPaths...)
+}
+
+var allowedFieldMaskPathGetters []func(string) []string
+
+// RegisterAllowedFieldMaskPathsGetter registers a function that is called with
+// the full RPC method when validating the field mask paths for an RPC.
+// If the function does not know the allowed field mask paths for the given RPC,
+// it should return nil.
+func RegisterAllowedFieldMaskPathsGetter(f func(rpcFullMethod string) []string) {
+	allowedFieldMaskPathGetters = append(allowedFieldMaskPathGetters, f)
+}
+
+func getAllowedFieldMaskPaths(rpcFullMethod string) []string {
+	if paths, ok := allowedFieldMaskPaths[rpcFullMethod]; ok {
+		return paths
+	}
+	for i := len(allowedFieldMaskPathGetters) - 1; i >= 0; i-- {
+		if paths := allowedFieldMaskPathGetters[i](rpcFullMethod); paths != nil {
+			return paths
+		}
+	}
+	return nil
+}
+
+type fieldMaskGetter interface {
+	GetFieldMask() types.FieldMask
+}
+
+var errForbiddenFieldMaskPaths = errors.DefineInvalidArgument("field_mask_paths", "forbidden path(s) in field mask", "forbidden_paths")
+
+func forbiddenPaths(requested []string, allowed ...string) (invalid []string) {
+nextRequested:
+	for _, requested := range requested {
```

Comment:
In my opinion this is ugly, error-prone and very confusing.
When I was looking through the fieldmask utility code in `ttnpb` I had to spend a lot of unnecessary effort to figure out which variables gets overloaded when and what do they evaluate to in different positions. IMO this should not happen and makes life of other developers harder.
Please avoid overloading variables like this(and other fieldmask utilities in `ttnpb`)

Reply:
Although I really don't see how this is confusing in such a small function, I renamed the variables for you.